syncs/shardedexpvar.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package syncs
+
+import (
+	"encoding/json"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sys/cpu"
+)
+
+// TODO(raggi): this implementation is still imperfect as it will still result
+// in cross CPU sharing periodically, we instead really want a per-CPU shard
+// key, but the limitations of calling platform code make reaching for even the
+// getcpu vdso very painful. See https://github.com/golang/go/issues/18802, and
+// hopefully one day we can replace with a primitive that falls out of that
+// work.
+
+// ShardedInt provides a sharded atomic int64 value that optimizes high
+// frequency (Mhz range and above) writes in highly parallel workloads.
+// The zero value is not safe for use; use |NewShardedInt|.
+// ShardedInt implements the expvar.Var interface.
+type ShardedInt struct {
+	shards  []intShard
+	nextGet atomic.Int64
+	// We absuse the mostly p-local and p-unique nature of the sync.Pool
+	// implementation to get a per-p shard that is on average likely to not
+	// share with other CPUs. The exact amount of sharing will vary based on the
+	// move frequency of the two schedulers and happenstance around the timing
+	// of pool borrows.
+	pool sync.Pool
+}
+
+type intShard struct {
+	atomic.Int64
+	_ cpu.CacheLinePad // avoid false sharing of neighboring shards
+}
+
+// NewShardedInt returns a new |ShardedInt|.
+func NewShardedInt() *ShardedInt {
+	m := &ShardedInt{
+		shards: make([]intShard, runtime.NumCPU()),
+	}
+	m.pool.New = func() interface{} {
+		return &m.shards[int(m.nextGet.Add(1)-1)%len(m.shards)]
+	}
+	return m
+}
+
+// Add adds delta to the value.
+func (m *ShardedInt) Add(delta int64) {
+	v := m.pool.Get().(*intShard)
+	v.Add(delta)
+	m.pool.Put(v)
+}
+
+// Value returns the current value.
+func (m *ShardedInt) Value() int64 {
+	var v int64
+	for i := range m.shards {
+		v += m.shards[i].Load()
+	}
+	return v
+}
+
+// GetDistribution returns the current value of each shard, this is intended for
+// observability/debugging only.
+func (m *ShardedInt) GetDistribution() []int64 {
+	v := make([]int64, len(m.shards))
+	for i := range m.shards {
+		v[i] = m.shards[i].Load()
+	}
+	return v
+}
+
+// String implements the expvar.Var interface
+func (m *ShardedInt) String() string {
+	v, _ := json.Marshal(m.Value())
+	return string(v)
+}
```

Comment:
IIRC the overall handler special cases some of the AppendFoo variants too, to have fewer allocs here

Reply:
It does, but looks like it's a private interface? Pretty sure this won't work 
https://github.com/golang/go/blob/95b433eed428afbb4ab32f0f2541774e939989c7/src/expvar/expvar.go#L48-L51
https://github.com/golang/go/blob/95b433eed428afbb4ab32f0f2541774e939989c7/src/expvar/expvar.go#L62-L68

I tried it out on the Go playground and yeah, looks like we can't make use of the special-case 😢  Maybe something to upstream?
https://go.dev/play/p/LIydisKTEDx

Reply:
Per later discussion added an implementation of the future encoding.TextAppender interface.